common/src/main/java/com/microsoft/identity/common/internal/activebrokerdiscovery/BrokerDiscoveryClient.kt

```diff
@@ -86,31 +90,42 @@ class BrokerDiscoveryClient(private val brokerCandidates: Set<BrokerData>,
         internal fun queryFromBroker(brokerCandidates: Set<BrokerData>,
                                      ipcStrategy: IIpcStrategy,
                                      isPackageInstalled: (BrokerData) -> Boolean): BrokerData? {
-            val methodTag = "$TAG:queryFromBroker"
 
             val installedCandidates = brokerCandidates.filter(isPackageInstalled)
 
-            installedCandidates.forEach { candidate ->
-                val operationBundle = BrokerOperationBundle(
-                    BrokerOperationBundle.Operation.BROKER_DISCOVERY_FROM_SDK,
-                    candidate.packageName,
-                    Bundle()
-                )
-
-                try {
-                    val result = ipcStrategy.communicateToBroker(operationBundle)
-                    return extractResult(result)
-                } catch (t: Throwable) {
-                    if (t is BrokerCommunicationException &&
-                        BrokerCommunicationException.Category.OPERATION_NOT_SUPPORTED_ON_SERVER_SIDE == t.category) {
-                        Logger.info(methodTag, "Tried broker discovery on ${candidate}. It doesn't support the operation")
-                    } else {
-                        Logger.error(methodTag, "Tried broker discovery on ${candidate}, get an error", t)
+            return runBlocking(dispatcher) {
```

Comment:
coroutine's mutex has to be invoked in a suspend function. that's why i wrapped it with runblocking.

Reply:
what I wanted to achieve is to provide synchronization across threads AND coroutines - and I picked coroutine's Mutex for that. 

Coroutine's mutex needs to be invoked _inside_ a suspend function, so i turned the whole block into one by wrapping it with runBlocking().

Not sure if that clarifies. Please let me know @p3dr0rv

Reply:
In this code, I have a method that do something like this.

```
A() {
    readCache()
    runBlocking {
         doWorkAsync()
    }
}
```

I originally put a java-thread-level-lock over A(). 

but let's say i invoke something like this.
```
runBlocking {
       launch {
             A(1)
       }
       launch {
             A(2)
       }
}
```

now both launch blocks will be executed on the main thread. let's say A(1) is invoked first. readCache() is done... then once it hits its runBlocking() - what I observed is is that A(1) will put doWorkAsync() at the end of the coroutine queue (can't find the official name here)

then A(2) will be executed "by a different coroutine, from the very same thread". 

If the lock is re-entrant, then A(2) would just breeze through readCache() like nothing happened....which is not an intended behavior.

If I changed the lock to a non-reentrant one (e.g. semaphore) it seems like doWorkAsync() will never be triggered. (since the lock is not yet released).